tests/beacon/state_machines/test_attestation_validation.py

```diff
@@ -0,0 +1,291 @@
+
+
+@pytest.mark.parametrize(
+    (
+        'attestation_latest_crosslink_root,'
+        'attestation_shard_block_root,'
+        'latest_crosslink_root,'
+        'is_valid,'
+    ),
+    [
+        (b'\x66' * 32, b'\x42' * 32, b'\x35' * 32, False),
+        (b'\x42' * 32, b'\x42' * 32, b'\x66' * 32, False),
+        (b'\x66' * 32, b'\x42' * 32, b'\x42' * 32, True),
+        (b'\x42' * 32, b'\x35' * 32, b'\x42' * 32, True),
+        (b'\x42' * 32, b'\x42' * 32, b'\x42' * 32, True),
+    ]
+)
+def test_validate_serenity_attestation_latest_crosslink_root(sample_attestation_data_params,
+                                                             attestation_latest_crosslink_root,
+                                                             attestation_shard_block_root,
+                                                             latest_crosslink_root,
+                                                             is_valid):
+    sample_attestation_data_params['latest_crosslink_root'] = attestation_latest_crosslink_root
+    sample_attestation_data_params['shard_block_root'] = attestation_shard_block_root
+    attestation_data = AttestationData(**sample_attestation_data_params).copy(
+        latest_crosslink_root=attestation_latest_crosslink_root,
+        shard_block_root=attestation_shard_block_root,
+    )
+
+    if is_valid:
+        validate_serenity_attestation_latest_crosslink_root(
+            attestation_data,
+            latest_crosslink_root,
+        )
+    else:
+        with pytest.raises(ValidationError):
+            validate_serenity_attestation_latest_crosslink_root(
+                attestation_data,
+                latest_crosslink_root,
+            )
+
+
+@pytest.mark.parametrize(
+    (
+        'attestation_shard_block_root,'
+        'is_valid,'
+    ),
+    [
+        (ZERO_HASH32, True),
+        (b'\x35' * 32, False),
+        (b'\x66' * 32, False),
+    ]
+)
+def test_validate_serenity_attestation_shard_block_root(sample_attestation_data_params,
+                                                        attestation_shard_block_root,
+                                                        is_valid):
+    attestation_data = AttestationData(**sample_attestation_data_params).copy(
+        shard_block_root=attestation_shard_block_root,
+    )
+
+    if is_valid:
+        validate_serenity_attestation_shard_block_root(
+            attestation_data,
+        )
+    else:
+        with pytest.raises(ValidationError):
+            validate_serenity_attestation_shard_block_root(
+                attestation_data,
+            )
+
+
+@settings(max_examples=1)
+@given(random=st.randoms())
+@pytest.mark.parametrize(
+    (
+        'num_validators,'
+        'epoch_length,'
+        'target_committee_size,'
+        'shard_count,'
+        'is_valid,'
+    ),
+    [
+        (10, 2, 2, 2, True),
+        (40, 4, 3, 5, True),
+        (20, 5, 3, 2, True),
+        (20, 5, 3, 2, False),
+        (10, 2, 3, 4, False),
```

Comment:
A bit confused here, it seems the only factor makes it's invalid is the `is_valid` parameter, not other system constants/configurations (`num_validators`, `epoch_length`, `target_committee_size`, `shard_count`). So you only need *one* `is_valid = False` case because it will be invalid anyway.

Reply:
True. I'll remove the last one